### PR TITLE
Implement basic version of sklearn estimator API

### PR DESCRIPTION
**Pull request by cle1109**
Addresses #2589. It is as simple as possible, but not simpler (I use `six` in order to keep the `sklearn` implementation of `set_params`).

### CODE REVIEW

File: mne/decoding/csp.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 import numpy as np
 from scipy import linalg
+import six
```
Comment (agramfort) [inline mne/decoding/csp.py]: use six from externals

File: mne/decoding/csp.py [hunk 2]
```diff
@@ -62,6 +63,40 @@ def __init__(self, n_components=4, reg=None, log=True):
...
+                sub_object.set_params(**{sub_name: value})
+            else:
+                # simple objects case
+                if key not in valid_params:
+                    raise ValueError('Invalid parameter %s for estimator %s. '
+                                     'Check the list of available parameters '
+                                     'with `estimator.get_params().keys()`.' %
+                                     (key, self.__class__.__name__))
+                setattr(self, key, value)
+        return self
```
Comment (agramfort) [inline mne/decoding/csp.py]: as it's really generic code I would put it in a mixin inherited class.

**Comment by cle1109** [comment]
@agramfort do you think we're ready to merge this one?

**Comment by agramfort** [comment]
there is no test

**Comment by cle1109** [comment]
Alright - what should we test here? I guess performing an actual grid search does not really qualify as a unit test.

**Comment by agramfort** [comment]
test in a Pipeline?

**Comment by cle1109** [comment]
A pipeline works without this modification (this is used in the CSP examples for instance). It's only if you perform a parameter search with CSP as part of a pipeline that the whole thing doesn't work anymore (because of the name mangling performed for each component and parameter in the pipeline).

**Comment by agramfort** [comment]
if you do

pipe = Pipeline([('csp', csp), ('svm', svm)]
pipe.set_params(csp__reg=0.2)

you have your test.

**Comment by cle1109** [comment]
I hope the rebase was OK...

**Comment by Eric89GXL** [comment]
Doesn't look right, there are unrelated commits :(

**Comment by cle1109** [comment]
Ouch. Any ideas how to quickly fix this mess?

**Comment by agramfort** [comment]
start a new clean branch from master and cherry pick the good commits.

**Comment by cle1109** [comment]
I think we're good again.

File: mne/decoding/tests/test_csp.py [hunk 3]
```diff
@@ -8,6 +8,8 @@
 from nose.tools import assert_true, assert_raises
 import numpy as np
 from numpy.testing import assert_array_almost_equal
+from sklearn.svm import SVC
+from sklearn.pipeline import Pipeline
```
Comment (agramfort) [inline mne/decoding/tests/test_csp.py]: do a nested import. Otherwise you make sklearn a strong dependency.

**Comment by agramfort** [comment]
that certainly explains why travis is not happy.

**Comment by agramfort** [comment]
merged by rebase

thanks @cle1109

**Pull request closed without merging**